### ISSUE TITLE
mypy not raising error on inexhaustive case/match statements in strict mode

### ISSUE THREAD

**Issue by Don-Burns** [python/mypy#19136]
**Bug Report**

(A clear and concise description of what the bug is.)

**To Reproduce**
Gist: https://mypy-play.net/?mypy=latest&python=3.12&gist=61f3db46cdaf7782c6679322e8809f47&flags=strict%2Ccheck-untyped-defs
```python
import enum
import random


class Test(enum.Enum):
    A = 1
    B = 2


val = list(Test)[random.randint(0, 2)]
reveal_type(val)
match val:
    case Test.A:
        print("A")
    # case Test.B:
    #     print("B")

val2 = random.randint(0, 2)
reveal_type(val2)
match val2:
    case 1:
        print("1")
```

**Expected Behavior**

Expected both of the match statements to raise errors due to not exhaustively matching on the enum/int values

**Actual Behavior**

File is passed as ok when it should raise errors
Pyright is able to catch both cases in strict mode as expected in case this helps: [pyright playground link](https://pyright-play.net/?strict=true&code=JYWwDg9gTgLgBAUwHYFcQFgBQpKzlAQyQBMINMssBjAGwIGd64AVBemACmTQDoBRVCACUALixwJcAIJwAvHACM4yQCE5cAEyUKmAG4Ea6msHYdW7IQG1CJMjxvFgSTgAYANJqEBdLCAIwqAAs4fRoxTEk4KgYEFjYYHilwyMiwKCdOACIpTKFlCQBiKJi49h4VZMki1PTnDkyVXO1QjXUHOwcMjndPX38gkIMNSolo%2BliFEZS0rsyFJp0gA)

**Your Environment**

- Mypy version used: tested on 1.15.0, 1.14.1, 1.10.1
- Mypy command-line flags: `--strict`
- Python version used: 3.12 & 3.10

**Possibly Related Issues?**
Below are the most relevant issues I could spot, but don't seem to quite align with the case here to the best of my understanding
- https://github.com/python/mypy/issues/19081

**Comment by A5rocks** [comment]
mypy doesn't have this feature, you need to add `case _: typing.assert_never(val)` to get exhaustiveness checking.

**Comment by Don-Burns** [comment]
Ah apologies then, I could have sworn I had seen this working in the past, but must have just been from my IDE's checking rather than running mypy (VsCode with pylance for IDE)
Would be a great feature though!  Understand if this gets closed though

**Comment by A5rocks** [comment]
Yeah maybe opt-in (what mypy currently has) is worse than opt-out (what pyright currently has) for strict mode at least. But I'm not too good with knowing what features mypy should add!

**Comment by Don-Burns** [comment]
If the maintainers are open to it, I can try to take a stab at implementing this, but admittedly I am not very familiar with mypy's code base to know how hard/easy this may be to tackle.

**Comment by A5rocks** [comment]
Implementation wise, I think this is simple: at the end of this loop:

https://github.com/python/mypy/blob/a8ec8939ce5a8ba332ec428bec8c4b7ef8c42344/mypy/checker.py#L5455

check whether mypy is in strict mode (maybe add a new mode?) and whether the narrowed type (copy how `current_subject_type` is done, I think?) is `UninhabitedType`.

**Comment by sterliakov** [comment]
It should be simple to add, but I don't think it's really a sensible check to enable globally. Python's match statement is not as good as Rust's (mostly because it's a statement and not a statement, so has no "return value"). Enforcing that `match` is always exhaustive is equivalent to requiring that any `if` has a corresponding `else`, which isn't something commonly enforced in python code. There's a simple `case other: assert_never(other)` option to enforce exhaustiveness of every match stmt separately.
